File: backend/app/core/upgrades/effects/armor_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Armor2(UpgradeImplementation):
    id = 'armor_2'
    name = 'Hardened Shell'
    description = 'Reduce incoming damage by 6%'
    rarity = 'epic'
    type = 'effect'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        appliers.effects.register_effect('protection', {'amount': 0.06})

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/effects/armor_2.ts
import { Upgrade, type UpgradeDef, type DamageRef } from '../Upgrade'
import { RarityID, UpgradeTypeID } from '../../data/ID'

export const Armor2Def: UpgradeDef = {
  id: "armor_2",
  name: "Hardened Shell",
  description: "Reduce incoming damage by 6%",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.Effect,
  cost: 20,
  effect: "protection",
  effectValue: 0.06,
  stackable: true,
  maxStacks: 4,
}

export class Armor2 extends Upgrade {
  onApply(): void {}

  modifyPlayerHurt(damage: DamageRef): void {
    // Always let at least 1 damage through
    damage.amount = Math.max(1, damage.amount * (1 - this.def.effectValue!))
  }
}
